tests/test_transform_dead_code.py

```diff
@@ -0,0 +1,101 @@
+# (C) Copyright 2018- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+import pytest
+
+from conftest import available_frontends
+from loki import Subroutine, FindNodes, Conditional, Assignment
+from loki.transform import dead_code_elimination
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_transform_dead_code_conditional(frontend):
+    """
+    Test correct elimination of unreachable conditional branches.
+    """
+    fcode = """
+subroutine test_dead_code_conditional(a, b, flag)
+  real(kind=8), intent(inout) :: a, b
+  logical, intent(in) :: flag
+
+  if (flag) then
+    if (1 == 6) then
+      a = a + b
+    else
+      b = b + 2.0
+    end if
+
+    if (2 == 2) then
+      b = b + a
+    else
+      a = a + 3.0
+    end if
+
+    if (1 == 2) then
+      b = b + a
+    elseif (3 == 3) then
+      a = a + b
+    else
+      a = a + 6.0
+    end if
+
+  end if
+end subroutine test_dead_code_conditional
+"""
+    routine = Subroutine.from_source(fcode, frontend=frontend)
+    # Please note that nested conditionals (elseif) counts as two
+    assert len(FindNodes(Conditional).visit(routine.body)) == 5
+    assert len(FindNodes(Assignment).visit(routine.body)) == 7
+
+    dead_code_elimination(routine)
+
+    conditionals = FindNodes(Conditional).visit(routine.body)
+    assert len(conditionals) == 1
+    assert conditionals[0].condition == 'flag'
+    assigns = FindNodes(Assignment).visit(routine.body)
+    assert len(assigns) == 3
+    assert assigns[0].lhs == 'b' and assigns[0].rhs == 'b + 2.0'
+    assert assigns[1].lhs == 'b' and assigns[1].rhs == 'b + a'
+    assert assigns[2].lhs == 'a' and assigns[2].rhs == 'a + b'
```

Comment:
```suggestion

    if (a > 2.0) then
      a = a + 5.0
    elseif (2 == 3) then
      a = a + 3.0
    else
      a = a + 1.0
    endif
  end if
end subroutine test_dead_code_conditional
"""
    routine = Subroutine.from_source(fcode, frontend=frontend)
    # Please note that nested conditionals (elseif) counts as two
    assert len(FindNodes(Conditional).visit(routine.body)) == 7
    assert len(FindNodes(Assignment).visit(routine.body)) == 10

    dead_code_elimination(routine)

    conditionals = FindNodes(Conditional).visit(routine.body)
    assert len(conditionals) == 2
    assert conditionals[0].condition == 'flag'
    assert not conditionals[0].has_elseif
    assert conditionals[1].condition == 'a > 2.0'
    assert not conditionals[1].has_elseif
    assigns = FindNodes(Assignment).visit(routine.body)
    assert len(assigns) == 5
    assert assigns[0].lhs == 'b' and assigns[0].rhs == 'b + 2.0'
    assert assigns[1].lhs == 'b' and assigns[1].rhs == 'b + a'
    assert assigns[2].lhs == 'a' and assigns[2].rhs == 'a + b'
    assert assigns[3].lhs == 'a' and assigns[3].rhs == 'a + 5.0'
    assert assigns[4].lhs == 'a' and assigns[4].rhs == 'a + 1.0'
```

This fails as expected, because the `has_elseif` flag remains for the last conditional although the conditional nested into the else branch has vanished:

```
self = Conditional::

    def __post_init__(self):
        super().__post_init__()
        assert self.condition is not None

        if self.body is not None:
            assert isinstance(self.body, tuple)
            assert all(isinstance(c, Node) for c in self.body)  # pylint: disable=not-an-iterable

        if self.has_elseif:
            assert len(self.else_body) == 1
>           assert isinstance(self.else_body[0], Conditional)  # pylint: disable=unsubscriptable-object
E           AssertionError

loki/ir.py:610: AssertionError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[36] > /etc/ecmwf/nfs/dh1_home_a/nabr/loki/naml-routine-inline-pragma/loki/ir.py(610)__post_init__()
-> assert isinstance(self.else_body[0], Conditional)  # pylint: disable=unsubscriptable-object
   5 frames hidden (try 'help hidden_frames')
```

Reply:
Aha, I knew I forgot something! :wink: Thanks for pointing this out, I've adjusted the test now accordingly and fixed the bug.